Add tools to validate code snippets in Markdown docs

rdipardo opened the pull request:
Just an idea I floated in https://github.com/exercism/fsharp/pull/958#issue-699449850
This adds one new workflow item that scans frontend Markdown docs for ungrammatical code snippets.
The new CI script works something like this:

```sh-session
PS C:\> Get-Help -Name './check-code-snippets.ps1' -Examples

NAME
    check-code-snippets.ps1

SYNOPSIS
    Validate F# code snippets in concept exercise documentation.


    -------------------------- EXAMPLE 1 --------------------------

    PS C:\>./check-code-snippets.ps1
    Verifies all *.md files under the `concepts` and `exercises` directories

    -------------------------- EXAMPLE 2 --------------------------

    PS C:\>./check-code-snippets.ps1 exercises
    Verifies the *.md files under the `exercises` directory only

    -------------------------- EXAMPLE 3 --------------------------

    PS C:\>./check-code-snippets.ps1 -Dirs "./concepts/recursion ./exercises/concept/pizza-pricing"
    Verifies the *.md files under `concepts/recursion` and `exercises/concept/pizza-pricing`
```

It's only job is running a small console app that does all the parsing and error reporting, with help from [`FSharp.Compiler.Service`]. 

The app gets compiled from source inside a new `tools-extra` directory; that seemed a better alternative to populating `bin`, where it would've been invisible without hacking `.gitignore`

[`FSharp.Compiler.Service`]: https://www.nuget.org/packages/FSharp.Compiler.Service

ErikSchierboom reviewed (commented):
First of all: great job implementing this. My main question would be: could we still have invalid (non-compiling) code? In some cases this might help demonstrate things.

rdipardo commented:
- rebased with upstream and synced workflow with renamed concepts
- moved the checker's source files to `tools/`
- delayed error message reporting until the end of each run so there's no need to scroll back the output: https://github.com/exercism/fsharp/pull/963/commits/f193e6e6e0bc4a65c656557bc95288b1503fcf7f#diff-6c373e7648d99eaca3425c2a31bc69850de2e88ae5524a0d92285b9604d264d1R43-R47
- made sure ignored paths with Unix separators are converted to native style so the PowerShell examples work on Windows without tweaking: https://github.com/exercism/fsharp/pull/963/commits/f193e6e6e0bc4a65c656557bc95288b1503fcf7f#diff-6c373e7648d99eaca3425c2a31bc69850de2e88ae5524a0d92285b9604d264d1R35 (*)

---
(*) There's one undocumented quirk with passing a list of paths in "command" mode (as distinct from the PS prompt); the former expects a comma-seperated string ("first/path,next/path"); at the prompt only a space-seperated string will work ("path/1 path/2"). 
But that's a feature of PowerShell, not the script.

ErikSchierboom commented on tools/CodeFenceChecker/Collector.fs:
```diff
@@ -0,0 +1,65 @@
+    let sep = Path.DirectorySeparatorChar
+    let isEmptyLine (line: string) = String.IsNullOrEmpty(line.Trim())
+
+    let private matchBlockStart line = Regex.IsMatch(line, @"^.*[`~]{3,}(fsharp)")
+    let private matchBlockEnd line = Regex.IsMatch(line, @"^.*[`~]{3,}$")
```
At some in the future, if you'd be interested, you could consider using parsing the Markdown using a library like [Markdig](https://github.com/xoofx/markdig) to find the code fence blocks (one can use an arbitrary number of backticks).

ErikSchierboom commented:
Merged! Fantastic work @rdipardo.
